File: 卡码/代码随想录/JS/动态规划/04-股票问题/LeetCode_0123.js
//123. 买卖股票的最佳时机 III
const maxProfit = function(prices) {
	const dp=[-prices[0],0,-prices[0],0];
	for(let i=1;i<prices.length;i++){
		dp[0]=Math.max(dp[0],-prices[i]);
		dp[1]=Math.max(dp[1],dp[0]+prices[i]);
		dp[2]=Math.max(dp[2],dp[1]-prices[i]);
		dp[3]=Math.max(dp[3],dp[2]+prices[i]);
	}
	return dp[3];
};

File: 卡码/代码随想录/Go/动态规划/04-股票问题/LeetCode_0123.go
// 123. 买卖股票的最佳时机 III
package main

import (
	"math"
)

func maxProfit(prices []int) int {
	dp := []int{-prices[0], 0, -prices[0], 0}
	for i := 1; i < len(prices); i++ {
		dp[0] = int(math.Max(float64(dp[0]), float64(-prices[i])))
		dp[1] = int(math.Max(float64(dp[1]), float64(dp[0]+prices[i])))
		dp[2] = int(math.Max(float64(dp[2]), float64(dp[1]-prices[i])))
		dp[3] = int(math.Max(float64(dp[3]), float64(dp[2]+prices[i])))
	}
	return dp[3]
}
